instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt

```diff
@@ -113,11 +116,149 @@ class RouteAlternativesTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::cla
+        firstAlternative.unregister()
+        assertTrue(firstAlternative.calledOnMainThread)
+
+        val countDownLatch = CountDownLatch(1)
+        runOnMainSync {
+            mapboxNavigation.historyRecorder.stopRecording {
+                logE("history path=$it", "DEBUG")
+                countDownLatch.countDown()
+            }
+        }
+        countDownLatch.await()
+
         // Verify alternative routes events were triggered.
         assertEquals(2, routes.size)
         assertTrue(mapboxNavigation.getRoutes().isNotEmpty())
         assertNotNull(firstAlternative.routeProgress)
         assertNotNull(firstAlternative.alternatives)
+
+        firstAlternative.alternatives!!.forEach { alternativeRoute ->
+            routes.forEach { originalRoutes ->
+                assertNotEquals(originalRoutes, alternativeRoute)
+            }
+        }
+    }
+
+    /**
+     * The test verifies that if set routes to Navigation that are come from alternatives + n
+     * (where n >= 1) external routes, alternatives routes observer is not triggered with these routes.
+     * That was happening in legacy versions of NN.
+     */
+    @Test
+    fun additional_alternative_is_not_force_to_invoke_alternatives_observer() {
+        // Prepare with alternative routes.
+        setupMockRequestHandlers(coordinates)
+        val primaryRouteOnly = listOf(requestDirectionsRouteSync(coordinates).first())
+
+        // Play the slower alternative route.
+        runOnMainSync {
+            mapboxNavigation.historyRecorder.startRecording()
+            mockLocationReplayerRule.playRoute(primaryRouteOnly.first())
+            mapboxNavigation.startTripSession()
+        }
+
+        // Wait for enhanced locations to start and then set the routes.
+        val firstLocationIdlingResource = FirstLocationIdlingResource(mapboxNavigation)
+        firstLocationIdlingResource.firstLocationSync()
+        runOnMainSync {
+            mapboxNavigation.setRoutes(primaryRouteOnly)
+        }
+
+        // The alternative route is missed, so we expect the route to be removed.
+        val firstAlternative = RouteAlternativesIdlingResource(mapboxNavigation)
+        firstAlternative.register()
+        mapboxHistoryTestRule.stopRecordingOnCrash("alternatives failed") {
+            Espresso.onIdle()
+        }
+        firstAlternative.unregister()
+
+        assertNotNull(firstAlternative.alternatives)
+
+        val externalAlternatives = DirectionsResponse.fromJson(
+            readRawFileText(activity, R.raw.route_response_alternative_continue)
+        ).routes().also {
+            assertEquals(1, it.size)
+        }
+
+        val nextAlternativeObserver = RouteAlternativesIdlingResource(mapboxNavigation)
+
+        var setRoutes: List<DirectionsRoute> = emptyList()
+        runOnMainSync {
+            setRoutes = mutableListOf(
+                mapboxNavigation.getRoutes().first()
+            ) + firstAlternative.alternatives!! + externalAlternatives
```

Comment:
Is `firstAlternative.alternatives!!` something that was generated by NN? What do we expect to find in this list here?

Reply:
We actually don't really worry about `firstAlternative.alternatives!!`, in this test we need to mix NN alternatives with the external route(s) and set the list as alternatives to `MapboxNavigation#setRoutes`. This way we check that NN is not triggering native AlternativeObserver

Reply:
What would happen if we set:
1) primary route + NN alternatives;
2) primary route + external alternatives
Would the logic be different? Do you think it's worth testing?
I thought I've left a similar comment somewhere but I can't find it...

Reply:
> primary route + NN alternatives;

that's what actually verifies at `alternative_routes_observer_is_not_triggered_for_set_routes` 

>primary route + external alternatives

it's actually the same scenario: NN had an issue with a set of alternatives that didn't provide by them. So excluding an alternative, removing the whole alternative list, or adding a new one to a list is the same scenario.